Fix: Only include fonts effectively used in the final PDF document

yuyiz67 opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Some guidelines are provided like this, in HTML comments, to expedite the code review before merging your contribution.

First, please make sure that you have read the documentation page on fpdf2 development:
https://py-pdf.github.io/fpdf2/Development.html

If you're new to contributing to open source projects,
you might find this free video course helpful: http://kcd.im/pull-request
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->
Font Optimization: Only include fonts effectively used in the final PDF document. Fonts added via set_font() or add_page() but not actually used in the document are no longer included in the final output. This reduces file size and improves performance, especially for documents with many fallback fonts
 Fixes Issue [#1382](https://github.com/py-pdf/fpdf2/issues/1382) <!-- This will automatically close issue #0 once the PR is merged: https://help.github.com/en/articles/closing-issues-using-keywords -->

<!-- Have you done all of these things?  -->
**Checklist**:

<!-- To check an item, place an "x" in the box like so: "- [x] Item description"
     Add "N/A" to the end of each line that's irrelevant to your changes -->

- [x] A unit test is covering the code added / modified by this PR

- [x] This PR is ready to be merged <!-- In your opinion, can this be merged as soon as it's reviewed? Else, this can be turned into a Draft PR -->

- [x] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder

- [x] A mention of the change is present in `CHANGELOG.md`

<!-- Feel free to add additional comments, and to ask questions if some of those guidelines are unclear to you! -->

<!--
Once a PR is merged, maintainers will add your name to the contributors table in README.md.
If they forget, or you do not wish this to happen, please mention it.
-->

By submitting this pull request, I confirm that my contribution is made under the terms of the [GNU LGPL 3.0 license](https://github.com/py-pdf/fpdf2/blob/master/LICENSE).

yuyiz67 commented:
hi @Lucas-C thanks for your detailed instruction and this is the PR. I cannot add reviewer so I could only mention you here. Thanks

Lucas-C commented on CHANGELOG.md:
```diff
@@ -34,6 +34,7 @@ This can also be enabled programmatically with `warnings.simplefilter('default',
 ### Fixed
 * [`FPDF.write_html()`](https://py-pdf.github.io/fpdf2/fpdf/fpdf.html#fpdf.fpdf.FPDF.write_html): Fixed rendering of content following `<a>` tags; now correctly resets emphasis style post `</a>` tag: hyperlink styling contained within the tag authority. - [Issue #1311](https://github.com/py-pdf/fpdf2/issues/1311)
 * [FPDF.footer()](https://py-pdf.github.io/fpdf2/fpdf/fpdf.html#fpdf.fpdf.FPDF.footer) does not "leak" its text style to the [table of contents](https://py-pdf.github.io/fpdf2/DocumentOutlineAndTableOfContents.html#table-of-contents) anymore
+* Font Optimization: Only include fonts effectively used in the final PDF document. Fonts added via set_font() or add_page() but not actually used in the document are no longer included in the final output. This reduces file size and improves performance, especially for documents with many fallback fonts. - Issue #1382
```
```suggestion
* Font Optimization: Only include fonts effectively used in the final PDF document. Fonts added via set_font() or add_page() but not actually used in the document are no longer included in the final output. This reduces file size and improves performance, especially for documents with many fallback fonts. - [Issue #1382](https://github.com/py-pdf/fpdf2/issues/1382)
```

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -3427,6 +3415,7 @@ def _render_styled_text_line(
                         self._resource_catalog.add(
                             PDFResourceType.FONT, current_font.i, self.page
                         )
+                    current_font_is_added = True
```
Shoudn't that be reset to False at tome point?

What happens when a text fragment with a different font is parsed?
Is it correctly added to the content stream?

Lucas-C commented on reproduce.py:
```diff
@@ -0,0 +1 @@
+a
```
This script should be removed from the PR

Lucas-C commented on test/linearization.pdf:
This file should not be added.
The corresponding unit test fail has `@pytest.mark.xfail`, it should not produce a PDF reference file.

Lucas-C reviewed (changes requested):
There are several problems with modified documents:
* text disapeared in `test/barcodes/barcodes_code39.pdf`
* card sizes changed in `test/fonts/fonts_emoji_glyph.pdf`
* text disapeared in ` test/fonts/fonts_set_builtin_font.pdf `
* first text line in black disapeared in ` test/free_text_annotation_all_parameters.pdf `
* etc.

This must be fixed before merging this PR.

Lucas-C commented:
Could you rebase your branch on this repo `master` branch please?

Once you have done that, you could check PDF differences using `scripts/compare-changed-pdfs.py`:
https://py-pdf.github.io/fpdf2/Development.html#visually-comparing-all-pdf-reference-files-modified-on-a-branch

yuyiz67 commented:
@Lucas-C  thanks for your quick review. I have changed my code. 
Firstly, I extracted the font set code into function `set_font_size_for_page` 
My assumption about font is simple: In a page, if a <font,size> is set, following text will use the <font, size> until a new <font,size> comes. 
So I added a variable `font_size_set_for_page` to indicate whether current page has set font and size. 
When a new page is added, font_size_set_for_page should be reset to false, so it is only reset in `add_page()` 
I added `set_font_size_for_page` to text(), add_text_notation(), _render_styled_text_line() because text will be rendered in these functions. 
However, after I refresh all unit test pdfs. I noticed a lot of difference in `test_html`, a good example is `test_html_features` 
If you open `test/html/html_features.pdf` you will see the text disappeared after <a> tag and came back after <font size> tag. In `test/html/html_long_ol_bullets.pdf` also some text disappeared. It feels like <a> tag and <ul> tag have different logic and removed the font. 
Also some text disappeared in `test/html/html_blockquote_indent.pdf` and `test/cell_skew_text.pdf`
I think my assumption about font is incorrect. In a page, after a <font,size> is set, when some code is added, the font will disappear. But I don't know what exactly made the font gone. If I know, I could add `font_size_set_for_page = False` after that code. 
Do you have any idea or suggestions about debugging this issue? Thank you very much

yuyiz67 commented:
I added a fix but these files still miss text, let me try to fix them:
test/html/html_table_with_null_text_in_span_cell.pdf
test/html/html_table_with_only_tds.pdf
test/outline/footer_leaking_style_on_toc.pdf
test/outline/test_start_section_horizontal_alignment.pdf
test/outline/toc_without_font_style.pdf (wrong font size)
test/table/table_valign_per_row.pdf
test/table/table_with_fill_color_set_beforehand.pdf
test/table/table_with_multiline_cells_and_split_over_3_pages.pdf
test/table/table_with_multiline_cells_and_without_headings.pdf
test/table/table_with_no_headers_nor_horizontal_lines.pdf
test/table/table_with_qrcode.pdf
test/table/table_with_ttf_font.pdf
test/table/table_without_headings.pdf
test/template/flextemplate_wrapmode.pdf
test/template/template_wrapmode.pdf
test/text/multi_cell_table_unbreakable_with_split_only.pdf
test/text/text_modes.pdf
test/transparency.pdf

yuyiz67 commented:
Just confirmed all pdfs are the same as master branch. Let me double check all code is good.

yuyiz67 commented on fpdf/fpdf.py:
```diff
@@ -3867,12 +3875,14 @@ def _perform_page_break(self):
+        self.font_size_set_for_page = False
         self._push_local_stack()
         self.add_page(same=True)
+        self._pop_local_stack()
```
previously, the sequence looks quite confusing:
pop_gs -> push_local -> add_page -> push_gs -> pop_local 

so here I changed it to correct sequence:
pop_gs -> push_local -> add_page -> pop_local -> push_gs 

This makes code more readable
@Lucas-C

yuyiz67 commented:
@Lucas-C  I have checked the pdfs with the script you mentioned. I think the code looks much better now. 
Main logic is to reset `font_size_set_for_page` to False for all states in  `_perform_page_break`.

Lucas-C commented:
Thank you for all your work on this @yuyiz67 👍

I may not have the time to review until Friday,
but I'll definitely get back to you before next week 🙂

yuyiz67 commented:
@Lucas-C  Thanks for running workflow. I tried to run it locally and found error :

>❯ scripts/pdfchecker.py --print-aggregated-report
PDF Checker 2.5.1  Copyright 2018-2024 Datalogics, Inc. All Rights Reserved
# AGGREGATED REPORT #
Errors:
* Uses Base 14 fonts not embedded in document:: x341 - First 3 files: test/free_text_annotation.pdf, test/barcodes/barcodes_interleaved2of5.pdf, test/alias_nb_pages.pdf

I will fix all errors and let you know after that.

Lucas-C commented:
Sorry I misclicked "Approve" on the wrong PR 😅

The tests are passing, it's great!

I'll take some time to review this properly tomorrow.

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -335,6 +335,7 @@ def __init__(
         self.font_family = ""  # current font family
         # current font style (BOLD/ITALICS - does not handle UNDERLINE nor STRIKETHROUGH):
         self.font_style = ""
+        self.font_size_set_for_page = False  # current font and size are already added to current page contents with _out
```
```suggestion
        self._font_size_set_for_page = False  # current font and size are already added to current page contents with _out
```
Let's use a private attibute: no need for end-users to fiddle with that 🙂

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -2250,15 +2250,18 @@ def set_font_size(self, size):
-                )
+        self.font_size_set_for_page = False
+
+    def set_font_size_for_page(self):
```
```suggestion
    def _set_font_size_for_page(self):
```
Let's use a private method: no need for end-users to fiddle with that neither

Lucas-C commented on fpdf/graphics_state.py:
```diff
@@ -56,6 +56,7 @@ def __init__(self, *args, **kwargs):
                 nom_lift=0.2,
                 denom_lift=0.0,
                 text_shaping=None,
+                font_size_set_for_page=False,
```
I do not think that this attribute needs to be part of teh GraphicsState:
could you please try to remove it from this class?

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -3867,12 +3875,14 @@ def _perform_page_break(self):
-            self._start_local_context(**prev_gs)
             self._push_local_stack()
-        self._pop_local_stack()
+            prev_gs["font_size_set_for_page"] = False
```
Is this really needed?

Lucas-C commented:
Good job overall @yuyiz67!

Excellent work 👍

I made a few final suggestions & remarks,
but once you have answered them and rebased your branch,
I think I'll merge this PR 🙂

yuyiz67 commented:
@Lucas-C 
Thanks for your review. Sorry I was busy in this week. 
I see you mentioned "I do not think that this attribute needs to be part of the GraphicsState: could you please try to remove it from this class?"
Please allow me to clarify my thoughts: 
GraphicsState is the state of a block. And fpdf could have nested blocks. 
I think `_set_font_size_for_page` is a sign for a block, not for a page because in a page, it is possible that  `_set_font_size_for_page` is true for inner block but is false for outer block.
I added a unit test `test_nested_context_font_usage_after_page_break` in recent commit. 
When we call multi_cell in context D, its  `_set_font_size_for_page` is already True but it is false for Context A. 
That's why I think this attribute should be in GraphicsState. 
But if you think we have another approach to resolve this nested situation, please help let me know and I will change the code. Thank you for your review and thoughts!

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -3427,6 +3433,8 @@ def _render_styled_text_line(
                         self._resource_catalog.add(
                             PDFResourceType.FONT, current_font.i, self.page
                         )
+                else:
+                    self._set_font_size_for_page()
```
Shouldn't `self._set_font_size_for_page()` also be called in the block above, if the `if` condition is `True`?

Lucas-C replied:
Handled in [`1b48d3e` (#1387)](https://github.com/py-pdf/fpdf2/pull/1387/commits/1b48d3e4a2594b96aa6339daef0a6b364a5cbdfe)

Lucas-C commented on test/fonts/test_unused_fonts.py:
```diff
@@ -0,0 +1,198 @@
+def test_nested_context_font_usage_after_page_break(tmp_path):
+    pdf = FPDF()
+    pdf.add_page()
+    pdf.add_font("Roboto-Regular", style="", fname=HERE / "Roboto-Regular.ttf")
+    pdf.add_font("Roboto-BoldItalic", style="", fname=HERE / "Roboto-BoldItalic.TTF")
+    pdf.add_font("DejaVuSans", style="", fname=HERE / "DejaVuSans.ttf")
+    pdf.add_font("Garuda", style="", fname=HERE / "Garuda.ttf")
```
```suggestion
    pdf.add_font(fname=HERE / "Roboto-Regular.ttf")
    pdf.add_font(fname=HERE / "Roboto-BoldItalic.TTF", style="BI")
    pdf.add_font(fname=HERE / "DejaVuSans.ttf")
    pdf.add_font(fname=HERE / "Garuda.ttf")
```
Mostly stylistic suggestions (not specifying `style` if not needed, and the font name will be deducted from `fname` if not provided), but I think that `style="BI"` was also missing for `Roboto-BoldItalic`.

Lucas-C replied:
Handled in [`1b48d3e` (#1387)](https://github.com/py-pdf/fpdf2/pull/1387/commits/1b48d3e4a2594b96aa6339daef0a6b364a5cbdfe)

Lucas-C commented:
> @Lucas-C Thanks for your review. [...] That's why I think this attribute should be stored in GraphicsState. not only in fpdf.

You are perfectly right!
Good job in crafting this great unit test.
Indeed, it demonstrates that this attribute should probably best defined in the `GraphicsStateMixin` to solve this case.

I took the liberty to add a commit to your branch: [`1b48d3e` (#1387)](https://github.com/py-pdf/fpdf2/pull/1387/commits/1b48d3e4a2594b96aa6339daef0a6b364a5cbdfe)
I performed the following changes:
* renamed `_font_size_set_for_page` into `current_font_is_set_on_page`, to make it clear that is is closely linke to the `current_font` attribute
* split the `_set_font_size_for_page()` method into two distincts methods: `_ensure_font_is_set_for_page()` & `_set_font_size_for_page()`, and introduced a new optional `wrap_in_text_object` boolean parameter
* used the new `_set_font_for_page()` method in `_render_styled_text_line()`
* made a minor fixup regarding `font_style=''` handling in `_start_local_context()`
* slightly modified `test_nested_context_font_usage_after_page_break()` to insert a 3rd font usage on the 2nd page

Does this looks good to you @yuyiz67? 🙂

Lucas-C commented:
I also rebased your `reproduce` branch on this repo `master` branch.

It was a bit cumbersome because it contains 18 commits 😅
If we need to rebase it again, we should probably stash commits beforehand, to make it easier to rebase.

Lucas-C commented:
There are many failing tests 😢
```
= 405 failed, 970 passed, 3 skipped, 2 xfailed, 7 warnings in 222.79s (0:03:42) =
```
We need to figure out why...
Sadly I won't be available for the rest of week-end.

Either you can investigate the issue @yuyiz67, or I'll have a look at the problem next week.

Lucas-C commented:
There are now only 45 tests that we need to investigate:
```
= 45 failed, 1331 passed, 2 skipped, 2 xfailed, 8 warnings in 221.04s (0:03:41) =
```

yuyiz67 commented:
@Lucas-C 
Thanks. I am trying to debug the failure on `test_cell_markdown_bleeding` 
New file `cell_markdown_bleeding.pdf` is different from master. 
The second row `No Markdown` is using the previous line's font style.

Because in method `def _parse_chars(self, text: str, markdown: bool)` , 
markdown is False and  it runs this code and return the Fragment with current graphics state. 
```
                    if fragment_text:
                        yield Fragment(
                            fragment_text, self._get_current_graphics_state(), self.k
                        )
```
This method assumes "current_graphics_state" is always the default font. 

in recent change of `_set_font_for_page` 
```
        if wrap_in_text_object:
            sl = f"BT {sl} ET"
```
this row `f"BT {sl} ET"` changed the default font. 

One way to fix this issue is to save the default font somewhere and make `_parse_chars` use default font rather than current graphics state. Then we have more freedom to change current graphics state. 
Do you have any better idea about this issue? Thanks

yuyiz67 commented:
@Lucas-C 
About the bug I mentioned above, I created a commit to fix: 
https://github.com/yuyiz67/fpdf2/commits/attempt_fix_for_markdown_bug/
but I didn't commit to current branch because I am not quite sure whether it is correct.

To check the difference, I added verbose in assert_pdf_equal so it could show difference like `git diff`. 
https://github.com/yuyiz67/fpdf2/tree/verbose_diff
I don't know how you compare the difference. If you think this method is good, I can create another PR for this. 

Debugging is hard. Thanks for your help.

Lucas-C commented:
Hi @yuyiz67

Sorry for not getting back to you, I got a busy week and got a cold 😓

I'll try to get back to you next week.

yuyiz67 commented:
@Lucas-C Thanks for heads up. Not rush. I didn't have time either. Hope you feel better soon!

Lucas-C commented:
I plan to merge this PR and then perform a new release of `fpdf2`.
Ideally before the end of the month.
Would you like to review it beforehand @yuyiz67 & @andersonhc? 🙂

Lucas-C commented:
@allcontributors please add @yuyiz67 for code
